### PR TITLE
Add Discord action (webhook or bot token)

### LINKED ISSUE #83: Discord Action

### PR DESCRIPTION
**Pull request by Zeroji**
This includes commits from #85 which hasn't been merged yet.
Fixes #83 and expands on #85 by enabling users to have a `DiscordAction` with a custom bot token and channel ID, instead of just a webhook.

Both ways work just fine on my machine, but I haven't tried using fresh tokens, let me know if you encounter authentication issues.

### CODE REVIEW

**Comment by dynomite567** [comment]
Looking at this, it implements a Discord bot way better and efficiently than my PR, #86, with no new dependencies at all. Good stuff!

File: pastepwn/actions/discordaction.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    name = "DiscordAction"
+
+    def __init__(self, webhook=None, token=None, channel_id=None, custom_payload=None, template=None):
+        super().__init__()
+        self.logger = logging.getLogger(__name__)
+
+        self.webhook = webhook
+        if webhook is None:
+            if token is None or channel_id is None:
+                raise ValueError('Invalid arguments: requires either webhook or token+channel_id arguments')
```
Comment (d-Rickyy-b) [inline pastepwn/actions/discordaction.py]: 👏 very nice

File: pastepwn/actions/discordaction.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    def perform(self, paste, analyzer_name=None):
+        """Send a message via Discord to a specified channel, without checking for errors"""
+        r = Request()
+        if self.template is None:
+            text = "New paste matched by analyzer '{0}' - Link: {1}".format(analyzer_name, paste.full_url)
+        else:
+            paste_dict = paste.to_dict()
+            paste_dict["analyzer_name"] = analyzer_name
+            text = self.template.safe_substitute(DictWrapper(paste_dict))
```
Comment (d-Rickyy-b) [inline pastepwn/actions/discordaction.py]: I created a much better way on how to use templates in the most recent master branch. But no worries, I will fix this later on :)

Comment (Zeroji) [reply pastepwn/actions/discordaction.py]: I belive this is Mxrk's work, inspired from `TelegramAction` source code, so make the change there too :)

**Comment by d-Rickyy-b** [comment]
Tested webhook from @Mxrk and it works perfectly!

grafik

But sadly I can't manage to get the bot part working. You write that you have not used "fresh" tokens, which means that you had a bot which connected via a library first and then you used that token?

**Comment by Zeroji** [comment]
Yes, as someone in the gist you linked yesterday mentioned, some initialization might be required.

I will try to generate some tokens for myself and try again, if I get it to work I will add the gateway ping in `DiscordAction.__init__`, I think. There might be issues if the message is sent too long after the ping, if you think standard pastepwn usage might fall in that category I will add logic to call it from `perform`.

**Comment by d-Rickyy-b** [comment]
@Zeroji That would be epic. Maybe also check the status code returned from the post method and do a gateway ping again.
Looking forward to hear back from you.

**Comment by Zeroji** [comment]
@d-Rickyy-b I'll have to check status codes because apparently this is a "happens occasionally at random intervals" kind of thing. I can reproduce the issue with fresh tokens, unfortunately I can't establish a Gateway connection without adding websockets. I'll give this a small hour, and drop it if I can't put something together - the alternative will be to run the token through any "proper" discord lib at least once.

**Comment by Zeroji** [comment]
image
It works! Took me the entire hour I allocated to it, I hope I covered edge cases because I'm going to sleep :stuck_out_tongue: 

I had to add `websockets` as a dependency unfortunately, but I used version 7.0 (8.0 is for python 3.6+) and tried to make my code compatible with Python 3.4. I haven't been able to build it on my machine (openssl errors...) so you'll have to let me know if something is broken.

The connection to the Gateway is done only when the request fails, and only once per DiscordAction (this might need to change in the future, maybe time-based?) because spamming Identify payloads (1000 per 24h) will reset the bot tokens.

Again, let me know if it works!

**Comment by d-Rickyy-b** [comment]
Very nice!! Thanks for all the effort. ❤️ 

Sadly pastepwn is a multithreaded application and hence it runs into an error regarding a missing event loop.

```
RuntimeError: There is no current event loop in thread 'ActionHandler'.
Exception in thread ActionHandler:
Traceback (most recent call last):
  File "C:\Users\Rico\PythonProjects\pastepwn\pastepwn\core\actionhandler.py", line 63, in _start
    action.perform(paste, analyzer.identifier)
  File "C:\Users\Rico\PythonProjects\pastepwn\pastepwn\actions\discordaction.py", line 112, in perform
    self.initialize_gateway()
  File "C:\Users\Rico\PythonProjects\pastepwn\pastepwn\actions\discordaction.py", line 86, in initialize_gateway
    asyncio.get_event_loop().run_until_complete(self._identify(ws_url))
  File "C:\Users\Rico\AppData\Local\Programs\Python\Python36-32\lib\asyncio\events.py", line 694, in get_event_loop
    return get_event_loop_policy().get_event_loop()
  File "C:\Users\Rico\AppData\Local\Programs\Python\Python36-32\lib\asyncio\events.py", line 602, in get_event_loop
    % threading.current_thread().name)
RuntimeError: There is no current event loop in thread 'ActionHandler'.
```

So I just added 
```python
loop = asyncio.new_event_loop()
asyncio.set_event_loop(loop)
```

And it seems to work just fine 🎉 . I am not sure what kind of side effects this brings in.

Thank you so much. Gonna merge this.

**Pull request merged**